File: DataStructures/Trees/TernarySearchTree.py
"""
- Ternary Search Tree -

Every node has three children. left, middle, and right

Advantages:
- TST support sorting
- TST is better than hashing, especially for search misses
- Faster than hashmaps and more flexible than binary search trees
"""

class Node:
    def __init__(self, char):
        self.char = char
        self.left = None
        self.middle = None
        self.right = None
        self.value = None


class TST:
    def __init__(self):
        self.root = None
    
    def insert(self, key, value):
        self.root = self.insert_helper(self.root, key, value, 0)
    
    def insert_helper(self, node, key, value, index):
        ch = key[index]
        if node is None:
            node = Node(ch)
        if node.char < ch:
            node.right = self.insert_helper(node.right, key, value, index)
        elif node.char > ch:
            node.left = self.insert_helper(node.left, key, value, index)
        elif index < len(key) - 1:
            node.middle = self.insert_helper(node.middle, key, value, index + 1)
        else:
            node.value = value
        return node

    def get(self, key):
        node = self.get_helper(self.root, key, 0)
        return node.value if node else None
    
    def get_helper(self, node, key, index):
        ch = key[index]
        if node is None:
            return None
        if node.char < ch:
            return self.get_helper(node.right, key, index)
        elif node.char > ch:
            return self.get_helper(node.left, key, index)
        elif index < len(key) - 1:
            return self.get_helper(node.middle, key, index + 1)
        else:
            return node


tst = TST()
tst.insert('hi', 100)
tst.insert('high', 200)
tst.insert('bye', 400)
tst.insert('goodbye', 900)

print(tst.get('hi'))
print(tst.get('high'))
print(tst.get('hig'))
print(tst.get('bye'))
print(tst.get('goodbye'))
print(tst.get('good'))

File: DataStructures/Trees/TernarySearchTree.js
/* 
- Ternary Search Tree -

Every node has three children. left, middle, and right

Advantages:
- TST support sorting
- TST is better than hashing, especially for search misses
- Faster than hashmaps and more flexible than binary search trees
*/

class Node {
    constructor(char) {
        this.char = char
        this.left = null
        this.right = null
        this.middle = null
        this.value = null
    }
}

class TST {
    constructor() {
        this.root = null
    }

    insert(key, value) {
        this.root = this.insertHelper(this.root, key, value, 0)
    }

    insertHelper(node, key, value, index) {
        let ch = key[index]
        if (!node) {
            node = new Node(ch)
        }
        if (node.char > ch) {
            node.left = this.insertHelper(node.left, key, value, index)
        } else if (node.char < ch) {
            node.right = this.insertHelper(node.right, key, value, index)
        } else if (index < key.length - 1) {
            node.middle = this.insertHelper(node.middle, key, value, index + 1)
        } else {
            node.value = value
        }
        return node
    }

    get(key) {
        let node = this.getHelper(this.root, key, 0)
        return node ? node.value : null
    }

    getHelper(node, key, index) {
        let ch = key[index]
        if (!node) return null
        if (node.char > ch) {
            return this.getHelper(node.left, key, index)
        } else if (node.char < ch) {
            return this.getHelper(node.right, key, index)
        } else if (index < key.length - 1) {
            return this.getHelper(node.middle, key, index + 1)
        } else {
            return node
        }
    }
}

const tst = new TST()
tst.insert('hi', 100)
tst.insert('high', 200)
tst.insert('bye', 400)
tst.insert('goodbye', 900)

console.log(tst.get('hi'))
console.log(tst.get('high'))
console.log(tst.get('hig'))
console.log(tst.get('bye'))
console.log(tst.get('goodbye'))
console.log(tst.get('good'))